cpp/include/kvikio/remote_handle.hpp

```diff
@@ -0,0 +1,327 @@
+ * A common way to do this, is to set the environment variables:
+ * `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY`.
+ *
+ * Other relevant options are `AWS_DEFAULT_REGION` and `AWS_ENDPOINT_URL`, see
+ * <https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-envvars.html>.
+ */
+class RemoteHandle {
+ private:
+  std::string _bucket_name{};
+  std::string _object_name{};
+  std::size_t _nbytes{};
+
+ public:
+  // Use of a default constructed instance is undefined behavior.
+  RemoteHandle() noexcept = default;
+
+  /**
+   * @brief Construct from a bucket and object name.
+   *
+   * @param bucket_name Name of the bucket.
+   * @param object_name Name of the object.
+   */
+  RemoteHandle(std::string bucket_name, std::string object_name)
+    : _bucket_name(std::move(bucket_name)),
+      _object_name(std::move(object_name)),
+      _nbytes(detail::get_s3_file_size(_bucket_name, _object_name))
+  {
+  }
+
+  /**
+   * @brief Construct from a remote path such as "s3://<bucket>/<object>".
+   *
+   * @param remote_path Remote file path.
+   */
+  RemoteHandle(const std::string& remote_path)
+  {
+    auto [bucket_name, object_name] = detail::parse_s3_path(remote_path);
+    _bucket_name                    = std::move(bucket_name);
+    _object_name                    = std::move(object_name);
+    _nbytes                         = detail::get_s3_file_size(_bucket_name, _object_name);
+  }
+
+  /**
+   * @brief Get the file size.
+   *
+   * Note, this is very fast, no communication needed.
+   *
+   * @return The number of bytes.
+   */
+  [[nodiscard]] std::size_t nbytes() const { return _nbytes; }
+
+  /**
+   * @brief Read from remote source into host memory.
+   *
+   * @param buf Pointer to host memory.
+   * @param size Number of bytes to read.
+   * @param file_offset File offset in bytes.
+   * @return Number of bytes read.
+   */
+  std::size_t read_to_host(void* buf, std::size_t size, std::size_t file_offset = 0)
+  {
+    KVIKIO_NVTX_FUNC_RANGE("AWS S3 receive", size);
+    auto t0 = std::chrono::high_resolution_clock::now();
+
+    auto& default_context = detail::S3Context::default_context();
+    Aws::S3::Model::GetObjectRequest req;
+    req.SetBucket(_bucket_name.c_str());
+    req.SetKey(_object_name.c_str());
+    const std::string byte_range =
+      "bytes=" + std::to_string(file_offset) + "-" + std::to_string(file_offset + size - 1);
+    req.SetRange(byte_range.c_str());
+
+    // To write directly to `buf`, we register a "factory" that wraps a buffer as a output stream.
+    Aws::Utils::Stream::PreallocatedStreamBuf buf_stream(static_cast<unsigned char*>(buf), size);
+    req.SetResponseStreamFactory(
+      [&]() { return Aws::New<detail::BufferAsStream>("BufferAsStream", &buf_stream); });
+
+    Aws::S3::Model::GetObjectOutcome outcome = default_context.client().GetObject(req);
+    if (!outcome.IsSuccess()) {
+      const Aws::S3::S3Error& err = outcome.GetError();
```

Comment:
_nit:_ let's use east consts throughout in this file.

Reply:
Done, `remote_handle.hpp` is the first file in KvikIO to go east :)